### PR TITLE
Added Independencies to get_independencies

### LINKED ISSUE #523: add tests for get_independencies method in BayesianModel

### PR DESCRIPTION
**Pull request by kshitij10496**
- [x] added local independencies to `get_independencies`
- [ ] added tests

fixes #523 #525

### CODE REVIEW

File: pgmpy/models/BayesianModel.py [hunk 1]
```diff
@@ -442,6 +442,12 @@ def get_independencies(self, latex=False):
                     if independent_variables:
                         independencies.add_assertions([start, independent_variables,
                                                        observed])
+            local_independence = self.local_independencies(start)
+            common_assertions = set(local_independence.get_assertions())
+            & set(independencies.get_assertions())
```
Comment (khalibartan) [inline pgmpy/models/BayesianModel.py]: @kshitij10496 This is not a valid syntax. travis CI was failing because of this.

Comment (kshitij10496) [inline pgmpy/models/BayesianModel.py]: @khalibartan Thanks. I have made an amend

**Comment by ankurankan** [comment]
@kshitij10496 Let me know when this is ready for review. And also changes in #532 overlap with the changes in this PR. Why don't you close one ?

**Comment by kshitij10496** [comment]
@ankurankan I have closed #532 .
I would like to discuss on the `hash` function

**Comment by ankurankan** [comment]
@kshitij10496 What about the `hash` function?

**Comment by kshitij10496** [comment]
I wanted to discuss about the `hash` function in the `Independencies` class.
You had suggested in the other PR that we should use `str` instead.

**Comment by ankurankan** [comment]
@kshitij10496 Yes, I meant use the inbuilt `hash` function on some unique string. So, you will have to have a way of constructing a string so that it's unique for each different object and then use the `hash` function on that.

**Comment by kshitij10496** [comment]
I tried to hash the object using its `__str__` before hand.
But the hash would depend on the order of assertions in the `Independencies` object.

Hence, I used adding the hash of each assertion in the object.
I know, this might not be the best way to hash but the probability of collisions is very small.

Do consider it and any advice on how can we do better ?

**Comment by khalibartan** [comment]
@kshitij10496 Could you replace `_|_` in `__str__` method by unicode `u"\u27C2"` . It is unicode for perpendicular ;) . Also the hash function using addition has good chances of collision and `__str__` will depend on order. Since get_assertion returns a list we can sort that list and convert it into a tuple and apply a tuple hash on it (built in tuple hash). Or is there a catch with this hash method ? But I don't think there is .

**Comment by kshitij10496** [comment]
@khalibartan Sounds cool.
If @ankurankan has no problem with it, I would be happy to make the necessary changes.

**Comment by kshitij10496** [comment]
@khalibartan What are your thoughts on the `hash` function in the `Independencies` class ?

**Comment by khalibartan** [comment]
@kshitij10496 I think it is good, we have already discussed about this on issue . It is slow , apart from it there is no issue . If I find any better implementation will definitely tell you . 

Also can you take a look at the local_independencies in BayesianModel , since you are working on independecies class there might be some changes that we need in it .

**Comment by kshitij10496** [comment]
@khalibartan About the `hash` function:
Initially, I attempted that approach.
However, I was unable to `sort` the list of `independencies` of the obtained by the `get_assertions()` method of the `Independencies` object.

**Comment by khalibartan** [comment]
@kshitij10496 Yeah It would happen actually since, it is a list of IndependenceAssertion objects. To sort it you have to do a bit of hard work, by iterating over each assertion object and converting them to inbuilt data type that can be sorted (str maybe) and then apply the method .
But now method seems a little vague , but it is as good collision free as inbuilt hash .
I'm thinking of some new method but you can try this method for now .

**Comment by khalibartan** [comment]
Actually I couldn't understand why we need `__hash__` for independencies , since self.independencies is a list of IndependenceAssertion , and IndependenceAssertion objects are hashable(after adding `__hash__` method to it , so can't we directly convert self.independencies to a set ?

**Comment by kshitij10496** [comment]
@khalibartan Yeah.
We can do that obviously.
For this issue, we dont need to hash the `Independencies` object.
However, I think we should add a `hash` function for code completeness.

**Comment by khalibartan** [comment]
I doesn't really makes sense to me. If you want to do so , you can go with the method I have [explained](https://github.com/pgmpy/pgmpy/pull/536#issuecomment-166674959).

**Comment by kshitij10496** [comment]
Basically, our motive is to add `hash` function and make the `Independencies` object hashabe.
Along the way, we want to solve the related issues.
And not the other way round.

**Comment by khalibartan** [comment]
Okay, things could have worked without it . So simply you can return the hash value of the by converting self.independencies to a set.

**Comment by kshitij10496** [comment]
I am getting your point here.
Are you suggesting to hash a set ?

**Comment by khalibartan** [comment]
Hmm.. interesting. I always thought `sets` were **hashable** , but actually `frozensets` are, so frozensets instead.

**Comment by kshitij10496** [comment]
Exactly !
Hence, I find it better to add a `hash` function

**Comment by ankurankan** [comment]
@kshitij10496 @khalibartan What's the status of this PR ?

**Comment by kshitij10496** [comment]
@ankurankan If you are fine with the methods, I will add the necessary tests.

**Comment by ankurankan** [comment]
@kshitij10496 Yes, please add the tests.

**Comment by kshitij10496** [comment]
Cool.

I am adding them :)

**Comment by khalibartan** [comment]
@kshitij10496 Tests are failing please give it a look.

**Comment by khalibartan** [comment]
@kshitij10496 Can fix the failing test. I want to work with these changes .

**Comment by kshitij10496** [comment]
@khalibartan I am looking into them.

**Comment by kshitij10496** [comment]
I am not entirely clear why the tests are failing.
The code runs perfectly fine on the local.

**Comment by khalibartan** [comment]
Try creating a separate `conda-env` for pgmpy and then run the tests. They are falling. I pulled in you branch and ran nosetests they failed . I haven't gone through the changes but I believe that order of assertions might be cause of failures, but it shouldn't have happened .

**Comment by kshitij10496** [comment]
Yeah. 
The tests are failing locally
However, the result seems to be correct.

Even I think the tests are failing either because of the order of assertions or the order of items in `event` of the assertions.

**Comment by khalibartan** [comment]
@kshitij10496 I have send PR to resolve your merge conflicts . You can pull changes of `pgmpy-dev` into your branch and your tests will pass then . :)
I found a bug in `__eq__` method. It was taking order into account.

**Comment by ankurankan** [comment]
@kshitij10496 @khalibartan Don't merge dev into this branch. Rather rebase your commits over the current dev. This will make the git history linear and much more cleaner.

**Comment by khalibartan** [comment]
Okay . My bad.

**Comment by khalibartan** [comment]
@kshitij10496 are you still working on this ?

**Comment by kshitij10496** [comment]
@khalibartan Yeah
I will fix this by tomorrow

**Comment by khalibartan** [comment]
@kshitij10496  Can you add a new implementation of the things, in IndependenceAssertion class by converting events from sets to frozensets and then returning their hash value. I did opened an issue for this but closed it later #575 .

**Comment by kshitij10496** [comment]
@khalibartan Yeah. Sure
I read that issue

**Comment by khalibartan** [comment]
@kshitij10496 Can you fix the merge conflicts ? I want to work with is_iequivalence class in BayesianModel

**Comment by ankurankan** [comment]
@kshitij10496 Tests are failing even after resolving the merge conflicts. Can you check?

**Comment by khalibartan** [comment]
@kshitij10496 are you still working on it ?

**Comment by yashu-seth** [comment]
@kshitij10496 Are you still working on this? The PR has been around for some time now. If you are unable to work on this for now, kindly let us know.

**Comment by chrisittner** [comment]
Regarding the change in `get_independencies`: I think `get_independencies` currently tries to compute all independencies by checking active trails/d-seperation. If done correctly, this should already include any dependencies found by `local_dependencies`, d-sep is a more general method. Thus adding the local independencies for each variable should be redundant once the issues (#607) in the algorithm are fixed.

**Comment by khalibartan** [comment]
@ankurankan I think you can close this PR , it has been here for quite a long time.

**Pull request closed without merging**